docs/enhancements/image-handling-service.md

```diff
@@ -0,0 +1,171 @@
+this should be compatible with the existing `GET /clusters/{cluster_id}/downloads/image`
+assisted service endpoint. When a user makes a request to this endpoint the
+image service will fetch the ignition and ramdisk (if necessary) from the
+assisted service and stream the customized iso to the client using the
+appropriate template image as a base. This will involve adding an endpoint to
+assisted service to expose the ramdisk data.
+
+### User Stories
+
+#### Story 1
+
+As a user deploying and running assisted service, I need image serving to be
+horizontally scalable and to not impact performance of the assisted service API.
+
+#### Story 2
+
+An additional deployment and service will be managed for all deployment methods.
+This includes SaaS, operator, and CI (minikube). Whether an additional image is
+created is still an open question.
+
+### Implementation Details/Notes/Constraints
+
+#### Authentication and Authorization
+
+The image service will not need to implement authentication or authorization
+directly as it doesn't manage any user data. The user will provide a token to
+the image service and the image service will directly pass that token to
+assisted service on each call. Assisted service will validate the token before
+giving the image service any user information.
+
+#### Template Images
+
+Currently, when the assisted service starts, it downloads and caches each RHCOS
+image specified in the `OPENSHIFT_VERSIONS` environment variable. It also
+creates a minimal iso image for each version by removing the root filesystem.
+These "template" images are the base for creating the discovery iso.
+
+The image service will take on the responsibility of managing the RHCOS template
+images and creating the minimal ISO template images. This means that the image
+service will require persistent storage (when not using S3).
+
+#### Image Streaming
+
+The version of the image service described here would require editing the image
+while streaming the download rather than the current behavior which stores the
+customized image to be downloaded in a separate call. Allowing the image to be
+stored would complicate the proposal and is discussed in the "Alternatives"
+section.
+
+#### Assisted Service API Changes
+
+For the image service to create the iso it needs some information from the
+assisted service. In the minimal ISO case, it will need the ignition and, if the
+user has configured static networking, an additional ramdisk image. For the full
+ISO, only the ignition is needed.
+
+Today the ignition is created and uploded separately to S3 by the assisted
+service when the discovery image is generated. This proposal involves changing
+that behavior to render the ignition on-demand rather than uploading it every
+time some image information changes. An additional API for the minimal ISO
+ramdisk will behave in the same way.
+
+### Risks and Mitigations
+
+Relying strictly on streaming will increase the load on the image service in
```

Comment:
It seems to me like an improvement after the image-service exists, but I wonder if most requests for a discovery image will be occurring around the same time if some basic caching could further reduce the burden on the assisted-service.

Reply:
> some basic caching could further reduce the burden on the assisted-service

This is interesting, caching what? Ignitions?

The problem I was getting at here was that we're moving download traffic from S3 to our application (in this case, the image service) by removing the option for a user to download the image directly from S3.

Reply:
> This is interesting, caching what? Ignitions?

Yeah. I was thinking about ignitions and ramdisks...in my head I think we could greatly reduce the load on the assisted-service coming from the image-service by caching a few of the responses that we would get from the assisted-service for ignition/ramdisk.

> The problem I was getting at here

I missed that if that is something that needs to be clarified. Issue I raised still seems worthy of consideration although it may be outside the scope of this enhancement.

Reply:
> This is interesting, caching what? Ignitions?

Yeah. I was thinking about ignitions and ramdisks...in my head I think we could greatly reduce the load on the assisted-service coming from the image-service by caching a few of the responses that we would get from the assisted-service for ignition/ramdisk.

> The problem I was getting at here

I missed that if that is something that needs to be clarified. Issue I raised still seems worthy of consideration although it may be outside the scope of this enhancement.

Reply:
@mhrivnak mentioned the same caching idea [here](https://github.com/openshift/assisted-service/pull/2172#discussion_r665739191), but I think we would run into issues unless we have a way to validate that a token isn't revoked from the image service side.

Currently we query the database to check for a cluster's existence to determine the validity of a token, but we can't do that from the image service. Also, if we ever wanted to add any kind of authorization to the tokens we use for fetching images we would run into the same problem.

I think it's a good idea to consider, but probably a bit out of scope for the first pass on this.